06_gpu_and_ml/stable_lm/main.py

```diff
@@ -0,0 +1,322 @@
+import base64
+import os
+import time
+from pathlib import Path
+from typing import Any, Dict, Generator, List, Optional, Union
+
+from pydantic import BaseModel
+from typing_extensions import Annotated, Literal
+
+import modal
+
+with open("requirements.txt", "r") as f:
+    requirements = base64.b64encode(f.read().encode("utf-8")).decode("utf-8")
+
+
+def build_models():
+    import torch
+    from huggingface_hub import snapshot_download
+    from transformers import AutoModelForCausalLM, AutoTokenizer
+
+    model_path = snapshot_download(
+        "stabilityai/stablelm-tuned-alpha-7b",
+        ignore_patterns=["*.md"],
+    )
+    m = AutoModelForCausalLM.from_pretrained(
+        model_path,
+        torch_dtype=torch.float16,
+        device_map="auto",
+        local_files_only=True,
+    )
+    m.save_pretrained(
+        model_path, safe_serialization=True, max_shard_size="24GB"
+    )
+    tok = AutoTokenizer.from_pretrained(model_path)
+    tok.save_pretrained(model_path)
+    [p.unlink() for p in Path(model_path).rglob("*.bin")]  # type: ignore
+
+
+image = (
+    modal.Image.conda()
+    .apt_install("git", "software-properties-common", "wget")
+    .conda_install(
+        "cudatoolkit-dev=11.7",
+        "pytorch-cuda=11.7",
+        "rust=1.69.0",
```

Comment:
Is this for `hf-transfer`? That package is written in Rust but shouldn't need the toolchain installed because it has published wheels on PyPi: https://pypi.org/project/hf-transfer/#files. Or is `conda` ignoring those PyPi wheels...

Reply:
Good point we're not building from source since it's compiled in the wheel. I didn't try without it though and conda does respect the `pip` installs we have so removing this would make the image leaner.